nifi-commons/nifi-expression-language/src/main/java/org/apache/nifi/attribute/expression/language/evaluation/functions/FormatEvaluator.java

```diff
@@ -47,23 +71,27 @@ public FormatEvaluator(final DateEvaluator subject, final Evaluator<String> form
             return new StringQueryResult(null);
         }
 
-        final QueryResult<String> formatResult = format.evaluate(evaluationContext);
-        final String format = formatResult.getValue();
-        if (format == null) {
-            return null;
+        DateTimeFormatter dtf;
+        if (preparedFormatter != null) {
+            dtf = preparedFormatter;
+        } else {
+            final QueryResult<String> formatResult = format.evaluate(evaluationContext);
+            final String format = formatResult.getValue();
+            if (format == null) {
+                return null;
+            }
+            dtf = DateTimeFormatter.ofPattern(format, Locale.US);
         }
 
-        final SimpleDateFormat sdf = new SimpleDateFormat(format, Locale.US);
-
-        if(timeZone != null) {
+        if ((preparedFormatter == null || !preparedFormatterHasCorrectZone) && timeZone != null) {
             final QueryResult<String> tzResult = timeZone.evaluate(evaluationContext);
             final String tz = tzResult.getValue();
-            if(tz != null && TimeZone.getTimeZone(tz) != null) {
-                sdf.setTimeZone(TimeZone.getTimeZone(tz));
+            if(tz != null) {
+                dtf = dtf.withZone(ZoneId.of(tz));
             }
         }
 
-        return new StringQueryResult(sdf.format(subjectValue));
+        return new StringQueryResult(dtf.format(subjectValue.toInstant().atZone(ZoneId.systemDefault())));
```

Comment:
Converting the result using `ZoneId.systemDefault()` can result in different behavior based on the running system configuration, is this the best approach?  At minimum it is worth documenting the behavior.  Given that `SimpleDateFormat` returns a `Date` object that depends on the system time zone internally, this approach might make sense, but it is worth testing and documenting.

Reply:
It is one approach that will give reasonable result when user wont provide time zone as an argument of `format` function and it worked this way before the change. Actually it is documented here: https://nifi.apache.org/docs/nifi-docs/html/expression-language-guide.html#format in "The number will be evaluated using the local time zone unless specified in the second optional argument". But you've right that it is not visible in tests. I've added `TestQuer.testFormatUsesLocalTimeZoneUnlessIsSpecified` test case.